### DISCUSSION TITLE
Node attributes vs. Node features

### DISCUSSION THREAD

**Discussion by chenruduan** [e3nn/e3nn#334] [Q&A] [answered]
Thanks for the great and easy-to-use package! I have some practical questions about node attribute (data.z) vs. node feature (data.x).

I understand that a node attribute does not get updated during graph convolutions while a node feature does. But I still would like to know more about what is the most physically meaningful way to choose between and their practical performance.

1) If I have an element-type-based property, say atom mass, should I count it as an attribute or a feature?

2) If I have a atomic property (not a fixed value w.r.t to the element type, say partial charge, should I count it as an attribute or a feature?

3) For input irreps that are not scalar, for example "2x0e+2x1o+1x2e", should I just arrange an array according to the input sequence? In the case above, I have data.x=[s00, s01, v00, v01, v02, v10, v11, v12, a00, a01, a02, a03, a04]. e3nn can recognize that s00 and s01 correspond to 2 scalars, {v00, v01, v02} and {v10, v11, v12} correspond to two separate vectors, etc?

Thank you in advance for your help!

Best regards,
Chenru

**Comment by mariogeiger** [comment, marked as answer]
Hi,

1. I don't know the answer to this question. I would guess that atom mass would be great for an attribute (data.z)
2. I still don't know, very curious to hear your outcome if you try both
3. Exactly, if you doubt about data format one sanity check to do is to check for the equivariance of your model.

**Reply by chenruduan** [reply]
Thanks! I will test a bit regarding the first two Qs.

**Reply by chenruduan** [reply]
@mariogeiger I find some weird behavior when I try to test this out with the `Network` model at `e3nn.nn.models.gate_points_2101`. Could you point me to the original paper for this implementation? Thanks!

**Reply by mariogeiger** [reply]
What kind of weird behaviors?

The model is a mix of the models of 2 papers:

Tensor field network 
3d steerable CNN

**Reply by chenruduan** [reply]
Thanks, I found these two in the repo! 
I am doing very simple tests on aspirin geometries (100 train and 100 test).
When using `{node_inputs: OHE, node_attr: OHE}` and `{node_inputs: OHE, node_attr: None}`, I can get energy and forces MAE close to the nequip paper (The OHE is "3x0e" in this case because aspirin has C, O, and H).
However, when I used random node inputs`{node_inputs: torch.rand(21, 3), node_attr: OHE}` (trying to mess up the model), the model still gives a good performance (MAE of 0.1 kcal/mol in energy and 0.2 kcal/mol/A in forces, which are both twice as large compared to the first two cases). Just wondering why this would be the case since I would expect randomness in the outputs as well.

**Reply by chenruduan** [reply]
@mariogeiger NVM. The model maps the random node inputs to a near-constant node inputs at the first layer.
